enterprise/server/remote_execution/nbd/nbdclient/nbdclient.go

```diff
@@ -0,0 +1,299 @@
+func (d *ClientDevice) waitForReady(ctx context.Context) error {
+	ctx, cancel := context.WithTimeout(ctx, readyCheckTimeout)
+	defer cancel()
+
+	f, err := os.Open(fmt.Sprintf("/sys/block/nbd%d/size", d.device.Index))
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	for {
+		if _, err := f.Seek(0, io.SeekStart); err != nil {
+			return err
+		}
+		rsize, err := io.ReadAll(f)
+		if err != nil {
+			return err
+		}
+		size, err := strconv.ParseInt(strings.TrimSpace(string(rsize)), 10, 64)
+		if err != nil {
+			return err
+		}
+		if size > 0 {
+			return nil
+		}
+		select {
+		case <-time.After(readyCheckPollInterval):
+			continue
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+// Mount connects the NBD device and mounts it to the given directory path. The
+// directory must already exist.
+func (d *ClientDevice) Mount(path string) error {
+	if d.mountPath != "" {
+		return status.FailedPreconditionError("already mounted")
+	}
+	if err := d.createDevice(); err != nil {
+		return status.WrapError(err, "failed to create device")
+	}
+	fstype := filesystemTypeString(d.metadata.GetFilesystemType())
+	if err := syscall.Mount(d.DevicePath(), path, fstype, syscall.MS_RELATIME, "" /*=data*/); err != nil {
+		return status.InternalErrorf("failed to mount %s (type %s) to %s: %s", d.DevicePath(), fstype, path, err)
+	}
+	d.mountPath = path
+	log.Infof("Mounted %s (type %s) to %s", d.DevicePath(), d.metadata.GetName(), path)
+	return nil
+}
+
+// Unmount unmounts the device from its current mount path and disconnects the
+// device.
+func (d *ClientDevice) Unmount() error {
+	if d.mountPath == "" {
+		return status.FailedPreconditionError("not mounted")
+	}
+	if err := syscall.Unmount(d.mountPath, 0); err != nil {
+		return status.InternalErrorf("unmount %s (%s): %s", d.mountPath, d.metadata.GetName(), err)
+	}
+	log.Infof("Unmounted %s", d.mountPath)
+	d.mountPath = ""
+	// Also disconnect the device since the host may be swapping out the backing
+	// file, and it's simpler to disconnect and reconnect rather than try to
+	// reconfigure a connected device.
+	if err := d.device.Disconnect(); err != nil {
+		return status.InternalErrorf("disconnect %s (%s): %s", d.DevicePath(), d.metadata.GetName(), err)
+	}
+	log.Infof("Disconnected %s", d.DevicePath())
+	d.device = nil
+	return nil
+}
+
+func (d *ClientDevice) ReadAt(p []byte, off int64) (int, error) {
+	var res *nbdpb.ReadResponse
+	err := d.doWithRedial(func() error {
+		var err error
+		res, err = d.client.Read(d.ctx, &nbdpb.ReadRequest{
+			Name:   d.metadata.GetName(),
+			Offset: off,
+			Length: int64(len(p)),
+		})
+		return err
+	})
+	if err != nil {
+		fatalf("%s: read failed: %s", d.DevicePath(), err)
+	}
+	if len(res.Data) > len(p) {
+		fatalf(
+			"block device server returned too many bytes (0x%x) for read at offset=0x%x length=0x%x",
+			len(res.Data), len(p), off)
+	}
+	copy(p, res.Data)
+	return len(res.Data), nil
```

Comment:
should this be:

```
bytesCopied := copy(p, res.Data)
return bytesCopied, nil
```

Reply:
good call, that seems slightly more conventional. Done